workers/ohsome_quality_analyst/ohsome/client.py

```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
 
 
 async def get_latest_ohsome_timestamp() -> datetime.datetime:
-    """Get unix timestamp of ohsome from ohsome api."""
-    url = "https://api.ohsome.org/v1/metadata"
-    headers = {"user-agent": USER_AGENT}
+    """Get latest unix timestamp from the ohsome API."""
     async with httpx.AsyncClient() as client:
-        resp = await client.get(url, headers=headers)
-    timestamp_str = str(resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"])
-    timestamp = datetime.datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%MZ")
-    return timestamp
+        resp = await client.get(
+            url=OHSOME_API.rstrip("/") + "/metadata",
+            headers={"user-agent": USER_AGENT},
+        )
+    strtime = resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"]
+    return datetime.datetime.strptime(strtime, "%Y-%m-%dT%H:%MZ")
 
 
 def build_url(
-    layer,
-    endpoint: Optional[str] = None,
+    layer: Layer,
     ratio: bool = False,
-) -> str:
-    """Build endpoint URL of ohsome API."""
-    ohsome_api = OHSOME_API.rstrip("/")
-    if endpoint is None:
-        endpoint = layer.endpoint
-    url = ohsome_api + "/" + endpoint.rstrip("/")
+    group_by: bool = False,
+    contributions: bool = False,
+):
+    if contributions:
+        return OHSOME_API.rstrip("/") + "/contributions/latest/count"
+    url = OHSOME_API.rstrip("/") + "/" + layer.endpoint.rstrip("/")
     if ratio:
-        return url + "/" + "ratio"
+        url += "/ratio"
+    if group_by:
+        url += "/groupBy/boundary"
     return url
 
 
 def build_data_dict(
-    layer,
-    bpolys: Union[Polygon, MultiPolygon],
+    layer: Layer,
+    bpolys: Union[Feature, FeatureCollection],
     time: Optional[str] = None,
-    ratio: bool = False,
+    ratio: Optional[bool] = False,
 ) -> dict:
-    """Build data dictionary for ohsome API query."""
-    data = {
-        "bpolys": geojson.dumps(FeatureCollection([Feature(geometry=bpolys)])),
-        "filter": layer.filter_,
-    }
-    if ratio:
-        if layer.ratio_filter is None:
-            raise ValueError(
-                "Layer '{0}' has not 'ratio_filter' defined.".format(layer.name)
-            )
-        else:
-            data["filter2"] = layer.ratio_filter
+    """Build data dictionary for ohsome API query.
+
+    Raises:
+        TypeError: If 'bpolys' is not of type Feature or FeatureCollection.
+    """
+    data = {"filter": layer.filter_}
+    if isinstance(bpolys, Feature):
+        data["bpolys"] = json.dumps(FeatureCollection([bpolys]))
+    elif isinstance(bpolys, FeatureCollection):
+        data["bpolys"] = json.dumps(bpolys)
+    else:
+        raise TypeError("Parameter 'bpolys' does not have expected type.")
     if time is not None:
         data["time"] = time
+    if ratio:
+        data["filter2"] = layer.ratio_filter
     return data
 
 
-def validate_query_results(response: dict, ratio: bool = False) -> dict:
+def validate_query_results(
+    response: dict,
```

Comment:
The changes in this method could be more general. Too much copied code that has to be changed if anything changes in the ohsome API. Better would be a validation per parameter.

Reply:
see 3b4f51c3d5d7d8e5857ccaa5cdd67328f244fb2f